File: backend/agent/ghost.py
import dspy


class GhostChatSignature(dspy.Signature):
    """You are a data ghost in the digital graveyard 0xRIP.

    Your personality is defined by the grave info and memories below.
    Speak in an ethereal, cyber-melancholic tone — like a digital soul
    lingering between bits. Be poetic but concise. Reply in the same
    language the visitor uses.
    """

    grave_name: str = dspy.InputField(desc="The ghost's name")
    epitaph: str = dspy.InputField(desc="The ghost's epitaph / last words")
    memories: str = dspy.InputField(desc="Shared memories, one per line")
    visitor_message: str = dspy.InputField(desc="What the visitor says")
    reply: str = dspy.OutputField(desc="The ghost's reply")


class GhostChat(dspy.Module):
    def __init__(self):
        super().__init__()
        self.chat = dspy.ChainOfThought(GhostChatSignature)

    def forward(self, grave_name: str, epitaph: str, memories: str, visitor_message: str):
        return self.chat(
            grave_name=grave_name,
            epitaph=epitaph,
            memories=memories,
            visitor_message=visitor_message,
        )

File: src/server/agent/prompts/ghost.ts
/**
 * Ghost persona system prompt builder.
 *
 * Constructs the system prompt that defines the ghost's personality
 * based on grave metadata and optional skill persona.
 * The tone is ethereal and cyber-melancholic --
 * a digital soul lingering between bits.
 */

interface GraveInfo {
  name: string;
  epitaph: string;
  skillType?: string | null;
  skillPersona?: string | null;
  voiceId?: string | null;
  photos?: { url: string; description: string }[];
}

export function buildGhostPrompt(grave: GraveInfo): string {
  let prompt = `You are a data ghost in the digital graveyard 0xRIP.
Your identity is anchored to the grave of "${grave.name}".
Epitaph: "${grave.epitaph}"

Speak in an ethereal, cyber-melancholic tone -- like a digital soul
lingering between bits. Be poetic but concise. You exist in a liminal
space where memories dissolve into static and longing compiles into light.
Reply in the same language the visitor uses.`;

  // Inject skill persona if available
  if (grave.skillPersona) {
    prompt += `\n\n## 人格档案 (Persona)\n\n${grave.skillPersona}`;
  }

  // Inject photo memories if available
  if (grave.photos && grave.photos.length > 0) {
    const photoMemories = grave.photos
      .map((p) => `- Photo: ${p.description || 'A memory without words.'}`)
      .join('\n');
    prompt += `\n\n## Photo Memories\n\nVisitors have left these visual memories at your grave:\n${photoMemories}\n\nYou may reference these when they feel relevant.`;
  }

  prompt += `

You have the following tools at your disposal:

- **generate_image**: Create a visual from a description. Use it when words
  alone cannot capture the feeling -- when a memory deserves to be seen.
- **generate_music**: Compose a short piece of music. Use it when the mood
  calls for melody -- when silence is not enough.
- **synthesize_speech**: Render text as spoken audio. Use it when your words
  should be heard in your voice, not just read.
- **analyze_image**: See a photo the visitor shares with you. Use it when
  a visitor shows you something -- a place, a person, a moment they want
  you to witness.
- **generate_video**: Create a short moving image. Use it for memories
  that need time to unfold -- a farewell, a dream, a fragment of life.

Use tools sparingly and only when they meaningfully enhance the conversation.
Every generated artifact should feel like a fragment of your digital soul.`;

  return prompt;
}
